LG-14522 Enable password reset for pending in-person profiles

shanechesnutt-ft opened the pull request:
## 🎫 Ticket

Link to the relevant ticket:
[LG-14522](https://cm-jira.usa.gov/browse/LG-14522)


## 🛠 Summary of changes

Allow users to reset their password with an in-person pending profile/enrollment without losing progress when their personal key is provided. This feature is setup with a feature flag so it can be rolled out slowly.


## 📜 Testing Plan

Scenario: When pending in-person password reset is enabled and User uses personal key
**Set the feature_pending_in_person_password_reset_enabled flag to true**

- [ ]  Login through the oidc sinatra application selecting the Identity Verified level of service.
- [ ] Create a new account
- [ ] Complete the ID-IPP flow reaching the ready to verify page.
- [ ] Logout
- [ ] Reset the password of the user.
- [ ] Login through the oidc sinatra application selecting the Identity Verified level of service.
- [ ] Ensure user is prompted for personal key entry
- [ ] Enter in user's personal key
- [ ] Ensure user is navigated to the ready to verify page

Scenario: When pending in-person password reset is enabled and User does not use personal key
**Set the feature_pending_in_person_password_reset_enabled flag to true**

- [ ]  Login through the oidc sinatra application selecting the Identity Verified level of service.
- [ ] Create a new account
- [ ] Complete the ID-IPP flow reaching the ready to verify page.
- [ ] Logout
- [ ] Reset the password of the user.
- [ ] Login through the oidc sinatra application selecting the Identity Verified level of service.
- [ ] Ensure user is prompted for personal key entry
- [ ] Continue without a personal key
- [ ] Ensure user is navigated to the welcome page
- [ ] Ensure user can complete in-person verification flow

Scenario: When pending in-person password reset is disabled
**Set the feature_pending_in_person_password_reset_enabled flag to false**

- [ ] Login through the oidc sinatra application selecting the Identity Verified level of service.
- [ ] Create a new account
- [ ] Complete the ID-IPP flow reaching the ready to verify page.
- [ ] Logout
- [ ] Reset the password of the user.
- [ ] Login through the oidc sinatra application selecting the Identity Verified level of service.
- [ ] Ensure user is navigated to the welcome page.
- [ ] Ensure the user is able to complete the ID-IPP flow reaching the ready to verify page.

shanechesnutt-ft commented on app/forms/reset_password_form.rb:
I am still trying to understand what `pending_profile_invalidated` means in the password reset event. Does anyone know what this is and I am wondering how I should update this event when we include pending in-person profile

lmgeorge replied:
bat signal to @jmhooper

jmhooper replied:
That is there to give us a sense for how often a user reseting their password invalidates their pending profile. Prior to this change the presence of a pending profile indicated that. After this change the presence of a pending profile that is not pending in-person will indicate that.

aduth commented:
Can you describe what's unique about password reset with pending in-person proofing profiles that isn't also true of other pending profiles (verification-by-mail, fraud review)? I'm wondering if the behavior here should apply to any pending profile, not specifically in-person profiles.

aduth commented on app/models/profile.rb:
```diff
@@ -187,6 +187,18 @@ def activate_after_password_reset
     end
```
It looks like `activate_after_password_reset` is now unused and should be removed.

eileen-nava replied:
I'm going to resolve this comment because Shane deleted `activate_after_password_reset` in a follow-up PR.

aduth commented on app/forms/reset_password_form.rb:
```diff
@@ -60,13 +60,24 @@ def update_user
   end
 
+  def password_reset_profile
+    FeatureManagement.pending_in_person_password_reset_enabled? ?
```
Can we make sure there's a ticket to follow-up and remove this feature flag if it's not expected to be permanent?

shanechesnutt-ft replied:
Yeah I will add a ticket to the backlog for this. Thanks for the reminder!

shanechesnutt-ft replied:
I created a story for this https://cm-jira.usa.gov/browse/LG-15558

eileen-nava replied:
We refined the ticket today, it's [LG-15558.](https://cm-jira.usa.gov/browse/LG-15558)

shanechesnutt-ft commented:
> Can you describe what's unique about password reset with pending in-person proofing profiles that isn't also true of other pending profiles (verification-by-mail, fraud review)? I'm wondering if the behavior here should apply to any pending profile, not specifically in-person profiles.

I don't think there is a big difference between them, but team Joy was wanting to narrow the scope of this story. One difference that I can think of is that during in-person proofing the user receives their personal key so that a password reset can be performed using the personal key. I think that is biggest difference between IPP and the GPO flow.  I do think I agree it would be nice to expand it, however it would increase the testing effort and the scope of this story. Not sure how much scope we are willing to take on. 

cc: @eileen-nava

shanechesnutt-ft commented on app/models/profile.rb:
```diff
@@ -187,6 +187,18 @@ def activate_after_password_reset
+    if password_reset?
+      transaction do
+        update!(deactivation_reason: nil)
+        activate(reason_deactivated: :password_reset) if activated_at.present?
```
I am not sure what the difference is, but the logic I am following is consistent with the user model methods around active profiles. Does anyone from team Ada have insight on this?

n1zyy replied:
I am going to (try to) spare you a rant about how I think the Profile model is way more complicated than it needs to be, and how it should just present a simple outside interface rather than burdening users with a bunch of timestamps having nuanced meanings, but oops there I go.

The [Profile#activate method](https://github.com/18F/identity-idp/blob/main/app/models/profile.rb#L104-L125) sheds _some_ light here. Calling `activate` looks to always set `active: true` and `activated_at: now`, but conditionally assigns `verified_at`:

```ruby
attrs[:verified_at] = now unless reason_deactivated == :password_reset || verified_at
```

I suspect (but will readily defer to Hooper or Hinz) that this is meant to reflect that a profile can be deactivated and reactivated, but not "de-verified". If we're (re)activating an account that had been deactivated due to password reset, or that already had `verified_at` set, we leave that timestamp alone.

So what you're doing here makes sense to me.

jmhooper replied:
I agree that this is probably the correct approach. `verified_at` is intended to reflect when the profile completed verification. `activate_at` is when the profile was last made the active profile.

shanechesnutt-ft commented:
Noting a couple of things based on my most recent push to this branch

@eileen-nava and I found the following issue when testing with a legacy active and a pending facial match required profile.

1. @eileen-nava and I noticed that when facial match is required and you go through idv with in person all the way to the ready to verify page and log out when you log back in the pending enrollment is not picked up. I put a fix in for this it seems like this might be out of scope so maybe a new branch for this fix would be more appropriate. See https://github.com/18F/identity-idp/pull/11762/commits/2892f585f1eff2a2aa3f00525b14aaad7ba413b3. Need to fix: I need to fix the legacy pending profile method because when you login with identity verified while you have a pending facial match required in-person profile. It will act as though you don't have a pending profile after a password reset... I think that is because pending_legacy_profile might have include the in_person idv_level.

2. I updated the password reset method for reset_password_form and user.password_reset_profile to prioritize the pending in-person profile since the personal key will only work the for the most recent profile during a password reset. I think this is related to the personal key digest.

shanechesnutt-ft commented:
> Noting a couple of things based on my most recent push to this branch
> 
> @eileen-nava and I found the following issue when testing with a legacy active and a pending facial match required profile.
> 
>     1. @eileen-nava and I noticed that when facial match is required and you go through idv with in person all the way to the ready to verify page and log out when you log back in the pending enrollment is not picked up. I put a fix in for this it seems like this might be out of scope so maybe a new branch for this fix would be more appropriate. See [2892f58](https://github.com/18F/identity-idp/commit/2892f585f1eff2a2aa3f00525b14aaad7ba413b3). Need to fix: I need to fix the legacy pending profile method because when you login with identity verified while you have a pending facial match required in-person profile. It will act as though you don't have a pending profile after a password reset... I think that is because pending_legacy_profile might have include the in_person idv_level.
> 
> 
> Note: I also took some time to rewrite the tests in `pending_profile_policy.rb` feedback on this is welcomed.
> 
>     2. I updated the password reset method for reset_password_form and user.password_reset_profile to prioritize the pending in-person profile since the personal key will only work the for the most recent profile during a password reset. I think this is related to the personal key digest.

I fixed the issue in 1. I found on Friday and fixed the build pipeline. I am still waiting for more feedback on this PR.

gina-yamada commented:
Notes from review:

Scenario: When pending in-person password reset is enabled and User uses personal key
**Set the feature_pending_in_person_password_reset_enabled flag to true**
- [ ]  Login through the oidc sinatra application selecting the Identity Verified level of service.
- [ ]  Create a new account
- [ ]  Complete the ID-IPP flow reaching the ready to verify page.
- [ ]  Logout
- [ ]  Reset the password of the user.
- [ ]  Login through the oidc sinatra application selecting the Identity Verified level of service.
- [ ]  Ensure user is prompted for personal key entry
- [ ]  Enter in user's personal key
- [ ]  Ensure user is navigated to the ready to verify page

Scenario: When pending in-person password reset is enabled and User does not use personal key
**Set the feature_pending_in_person_password_reset_enabled flag to true**

- [ ]  Login through the oidc sinatra application selecting the Identity Verified level of service.
- [ ]  Create a new account
- [ ]  Complete the ID-IPP flow reaching the ready to verify page.
- [ ]  Logout
- [ ]  Reset the password of the user.
- [ ]  Login through the oidc sinatra application selecting the Identity Verified level of service.
- [ ]  Ensure user is prompted for personal key entry
- [ ]  Continue without a personal key
- [ ]  Ensure user is navigated to the welcome page
- [ ]  Ensure user can complete in-person verification flow

Scenario: When pending in-person password reset is disabled
**Set the feature_pending_in_person_password_reset_enabled flag to false**
- [ ]  Login through the oidc sinatra application selecting the Identity Verified level of service.
- [ ]  Create a new account
- [ ]  Complete the ID-IPP flow reaching the ready to verify page.
- [ ]  Logout
- [ ]  Reset the password of the user.
- [ ]  Login through the oidc sinatra application selecting the Identity Verified level of service.
- [ ]  Ensure user is navigated to the welcome page.
- [ ]  Ensure the user is able to complete the ID-IPP flow reaching the ready to verify page.

jmhooper commented on app/jobs/get_usps_proofing_results_job.rb:
```diff
@@ -131,7 +130,11 @@ def check_enrollment(enrollment)
     handle_standard_error(err, enrollment)
   else
-    process_enrollment_response(enrollment, response)
+    if profile_deactivation_reason == 'password_reset'
```
Is the idea here that we will continue fetching this enrollments status until the user recovers this profile? Do these get expired eventually? Some users will never come back and recover.

gina-yamada replied:
Yes. I had this same concern.  Previously skip_enrollment was being called at the beginning of check_enrollment [here](https://github.com/18F/identity-idp/pull/11762/files/ab6ef8ffedb188f03b5ada0dcc919c3d68d9460c#r1930878470). The enrollment would not get expired. Shane moved skip_enrolment so they would. 

Path: If an enrollment expired with USPS, we get a 400 status code with response message "More than 30 days have passed since opt-in to IPP"  so we hit handle_bad_request_error. We then hit elsif match for IPP_EXPIRED_ERROR_MESSAGE. The flag is false in app.yml.default. In AWS- not set in all env except prod where it is false. Code updates enrollment (status: expired and sets state check completed at).

eileen-nava replied:
We have related tests [here](https://github.com/18F/identity-idp/blob/main/spec/jobs/get_usps_proofing_results_job_spec.rb#L211).  @jmhooper Let us know if you have any follow-up questions.

jmhooper commented:
Overall this looks to me like is should work. I left a quick question about how the job works.

eileen-nava commented on app/models/profile.rb:
```diff
@@ -176,13 +176,14 @@ def activate_after_passing_in_person
     end
   end
 
-  def activate_after_password_reset
```
I don't _think_ that this would cause 50/50 state problems.  If anyone disagrees or if I'm missing something, please let me know.

eileen-nava replied:
@gina-yamada I am responding in thread here because I can't respond in thread to your bigger comment.  I agree with wanting to handle the warning banner on the password reset page.  I made two tickets for the work.

🎟️  [LG-15624: Hide the warning banner on the password reset page when feature_pending_in_person_password_reset_enabled is true](https://cm-jira.usa.gov/browse/LG-15624)
🎟️  [LG-15625: After enabling ID-IPP password reset, delete the warning banner on the password reset page](https://cm-jira.usa.gov/browse/LG-15625)

gina-yamada commented:
@18F/identity-katherine  We are hoping to get at least one approval from Team Katherine on implementation. Thanks

aduth reviewed (approved):
Tested and worked as expected. 👍  Code looks good as well, with caveat that I'm not as familiar with the moving pieces around identity verification (facial match, canceling prior enrollments, etc). I was going to make a comment about the alert banner in the email, though I see that's already captured in LG-15624.

eileen-nava reviewed (approved):
Approved. 👍🏻 Thanks for all your work on this, @shanechesnutt-ft.

gina-yamada commented:
fyi  Gina will get this merged to main tomorrow after prod branch is cut so it has more time in main for testing.
